blocks/hooks/autocompleters.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * External dependencies
+ */
+import { clone } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { addFilter, applyFilters } from '@wordpress/hooks';
+import { Component } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import defaultCompleters from '../autocompleters';
+
+/**
+ * Wrap the default Autocomplete component with one that
+ * supports a filter hook for customizing its list of autocompleters.
+ *
+ * Since they may be many Autocomplete instances at one time, this component
+ * applies the filter on demand, when the component is first focused after
+ * receiving a new list of completers.
+ *
+ * This function is exported directly for unit test.
+ *
+ * @param  {Function} Autocomplete Original component.
+ * @return {Function}              Wrapped component
+ */
+export function filterAutocomplete( Autocomplete ) {
```

Comment:
There is a convention to prefix all HOCs with `with` prefix. In this case, it might be better to name it `withFilteredAutocompleters`.